feat: migrate to pnpm for better workspace DX

ojeytonwilliams opened the pull request:
- feat: npm -> pnpm part 1
- fix: correct mocha path for curriculum tests
- fix: update npm ci -> pnpm install

This is still very much in draft mode.  So far I've done the bare minimum to get it up and running locally.  If it passes in CI I still need to:

- [x] replace all remaining `npm`s with `pnpm`s (all but Dockerfiles which I didn't want to touch)
- [x] update the docs with an short guide to installing pnpm (very short guide. More of a link, if I'm honest)
- [ ] other stuff I've forgotten

---

Assuming we like and want to use this, we'll need to update the pipelines before we can deploy again.

Assuming all goes well, the followups are:

- enable pnpm caching
- update Dockerfiles
- clean up the scripts a bit
- delete/update our custom code cypress install code

ojeytonwilliams committed:
feat: npm -> pnpm part 1

This resolves the issues with the gatsby client (gatsby-plugin-pnpm
deals with the fact that gatsby is relying on its own dependencies
being de-duped)
and challenge-editor (which doesn't seem to want to automatically install
codemirror and needed its own eslint config)

ojeytonwilliams committed:
fix: correct mocha path for curriculum tests

ojeytonwilliams committed:
fix: update npm ci -> pnpm install

Includes pnpm action where necessary

ojeytonwilliams committed:
fix: use select workspace with -F not -w

ojeytonwilliams committed:
fix: reorganise packages and restrict hoisting

pnpm works best if the workspaces keep their own dependencies, since
dependencies are not flattened and then what node resolves from a
require is predictable.

@types seem to be a special case and more care is required to prevent
them getting smushed together in the root (hence the .npmrc)

ojeytonwilliams committed:
fix: add types for tools + root

ojeytonwilliams committed:
fix: decouple challenge-auditor from client

ojeytonwilliams committed:
fix: add ui-components types

ojeytonwilliams committed:
fix(client): use the latest types for react 16

ojeytonwilliams committed:
fix: prettify

ojeytonwilliams committed:
fix: prettierignore pnpm-lock

ojeytonwilliams committed:
fix: relax hoisting

Turns out pnpm works just fine with types. I don't know what was going
wrong before, but there are no-longer any type conflicts.

ojeytonwilliams committed:
fix: add @redux-saga/core to fix eslint issue

It seems to only be redux-saga that import/named can't cope with, so it
is probably okay to work around this one.

ojeytonwilliams committed:
chore: add chai to tools/scripts/build

ojeytonwilliams committed:
fix: add store to root for cypress

ojeytonwilliams committed:
refactor: name pnpm setup actions

ojeytonwilliams committed:
fix: add pnpm setup to cypress web

ojeytonwilliams committed:
fix: add pnpm setup to node tests

ojeytonwilliams committed:
fix: allow cypress to download binaries

If we want to keep preventing cypress from downloading binaries, we can
figure out a workaround, but I'm allowing it to ease the transition to
pnpm.

My guess about why this is happening is that npm triggers Cypress's
postinstall script, but pnpm does not (because pnpm install only
installs if necessary, perferring to link)

ojeytonwilliams committed:
chore: npm run to pnpm run

ojeytonwilliams committed:
chore: re-enable pre/post scripts

ojeytonwilliams committed:
fix: update build scripts for client

ojeytonwilliams committed:
fix: use pnpm for e2e-mobile

ojeytonwilliams committed:
fix: npm -> pnpm in node scripts

ojeytonwilliams committed:
chore: update log

ojeytonwilliams commented:
I'll update the docs and engines next, but everything else is ready for review.

If you don't have `pnpm` then `npm install -g pnpm` works just fine, but the install guides are [here](https://pnpm.io/installation) if you're curious.

raisedadead reviewed (dismissed):
You should separate the TypeScript changes from pnpm onboarding/migration. Those should be taken in regardless if we move or not.

Sboonny commented on client/src/components/SolutionViewer/SolutionViewer.tsx:
```diff
@@ -9,7 +9,7 @@ type Props = {
 type Solution = Pick<ChallengeFile, 'ext' | 'contents' | 'fileKey'>;
 
-function SolutionViewer({ challengeFiles, solution }: Props) {
+function SolutionViewer({ challengeFiles, solution }: Props): JSX.Element {
```
Sorry to say it after you are done, but can we disable the rule that force to explicit the return type of the function.

If we aren't interacting with contributors that don't know if we add this to please eslint, or the type of the return function is actually import, then think that the explicit type is indicating an issue with their code, may result in weird code to work around the function.

Sboonny replied:
A good example is this file 

https://github.com/freeCodeCamp/freeCodeCamp/blob/53164e71f160ac156de2706b77c28ae18e9ca027/client/src/components/formHelpers/form.tsx#L108-L116

Although it looks fine, in the place where `DynamicForm` is used, `hidebutton` is always `false`, and `buttonText` is always true.

So people are using if statements to work around the type checkers, while we don't need to

ShaunSHamilton replied:
For the most part, we should take advantage of the fact that we use TypeScript; we should use the strictest rules we can. The explicit returns for functions is a very useful rule.

I would suggest we as reviewers/maintainers be proactive with correcting/guiding Campers who submit code with "weird" code. For example, I have a red flag go off in my brain when I see any type assertion, and take time to see why that assertion is there, and if it is even necessary.

Sboonny replied:
if this is Rust, I will be all over this and in favor of this, but in TypeScript it has "feature", which makes the return type of function overwrite the infered typed, and not only that if you create a function that except `foo`, and add this code:

```ts
const something = (): randomType => { return foo as const}
```

Can you expect the type of returned value of `something`? it's `randomType` although you explicitly requested it return foo type in the return function, this is way we shouldn't use "feature", that to be honest broken, and use the normal ways to return the function values when needed.

Sboonny replied:
TD;LR if TypeScript doesn't have "feature" that overwrite the inferred return type or the explicit type, we should allow this.

naomi-lgbt replied:
<img width="246" alt="image" src="https://user-images.githubusercontent.com/63889819/218272254-d085fed7-3b99-4717-b265-f8a5b29bb1c9.png">

That code would throw a compiler error, because your returned value doesn't match the function signature. And that's exactly why explicit return types are important - it ensures that what we are returning is actually what we want to return, rather than a potential developer error.

ShaunSHamilton replied:
As Naomi said, the whole point of the explicit return is developers **have** to know what they want to return from a function. So, anything that _"accidentally"_ does not return the declared type should throw a type error:

<details>
  <summary>Example</summary>

```typescript
// I know I want a function that always returns a string
function getMatch(): string {
    const str = "a";
    const match = str.match(/b/)?.[0];
    return match;
}
```

Yields:

```bash
Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.
```

</details>

> if this is Rust, I will be all over this and in favor of this

Me too 😁

Sboonny replied:
> And that's exactly why explicit return types are important

And that's exactly why explicit return types are **dangerous**, because when someone consumes the functional the message they receive is viewed as a source of truth, but it isn't the truth because TypeScript can and do lie, which is **dangerous** especially in a codebase that half of it isn't TypeScript, and the many types in the other half are typed wrong.

We are asking maintainers and contributors to always return something correct, when they produce a function, but this isn't realistic. 

This is why when we as maintainers know that the function produced is important, we should explicitly put the return of the produced function `as const`, so when it's consumed, the types of the consumer should align with the `as const`, otherwise we get type error in consumption, and the type error will tell directly which part is the wrong one, the production or the consumption.

Sboonny replied:
You know what, let's decide a middle ground. this eslint rule can be applied in the API/Backend freely, but it shouldn't be applied in the client side.

If this is accepted, I am happy.

ShaunSHamilton replied:
We can discuss this in the next dev meeting.

> We are asking maintainers and contributors to always return something correct, when they produce a function, but this isn't realistic.

This is the whole point of TypeScript, though. Say what you expect in, what should come out, and complain when those conditions are not met.

ojeytonwilliams replied:
> And that's exactly why explicit return types are dangerous, because when someone consumes the function, the message they receive from TypeScript is viewed as a source of truth, but it isn't the truth because TypeScript can and do lie, which is dangerous especially in a codebase that half of it isn't TypeScript, and the many types in the other half are typed wrong.

You can 'trick' TypeScript via double type assertions like

```ts
const fakeNumber = 'pretending to be a number';
const notReallyANumber: number = fakeNumber as unknown as number;
```

but this isn't what this rule is concerned with. It requires us to set the return type and that's perfectly safe.

```ts
function okay(): string | number {
  return 'just a string'
}

function notOkay(): string {
  let ambiguous: string | number = Math.random() < 0.5 ? '': 0;
  return ambiguous
}
```
This a little contrived, but the `okay` function is fine because it's always safe to make a type _less_ specific. You can't get runtime errors this way, since you have to check if the returned data is a string or a number before using it.

That's not true for `notOkay`, since it really could be a string or a number. You are not allowed to narrow the return type - it's a compile-time error.

In both cases TS is giving us type safety, so the question is just "is this helpful?" not "is this dangerous?".

I can attest to it being a bit of a chore, but it does encourage the creation of more readable types. Also, if you're the one creating or updating the functions it should be less of a chore since you (should) understand the change.

naomi-lgbt replied:
In a codebase as large as ours, I'm a firm believer that explicit return types help a *lot* more than they hinder.

Sboonny replied:
I know that it's hard for me to convey what I mean, and this is prolonged than needed, so you can ignore this to keep things going, but:

- I don't appreciate the concept of repeating the same example, as I don't know it. I do know what you mean, and I think it is not only not worth it, but it's dangerous too, because we with the "big code base" and not bloated one, that define same values multiple times and the same component multiple times, barely know if the types are correct.

```ts
const fakeNumber = 'pretending to be a number';
const notReallyANumber: number = fakeNumber as unknown as number;
```

- This example is flawed in itself, because I didn't bring "tricking" TypeScript, but I brought that TypeScript can lie to use about the return result, and will lie to us.

---

And as I said, if you guys want to apply it, apply it in API and leave the /client away from it. 

If you don't want to do that, just don't do it, but don't bring the concept of "tricking" TypeScript, because this isn't about it, and don't mention that it's helpful for "big" codebase, because it really isn't, the "big" codebase doesn't need another rail guard, that can break and does lie.

ojeytonwilliams replied:
It's fine to discuss this here.

> TypeScript can lie to us about the return result, and will lie to us.

Rather than talking in the abstract, we should be concrete. Could you point to some code that has the problem you're concerned about? To be specific: cases where you have a function

```ts
function name() {
// ...
}
```
and adding an explicit type like

```ts
function name(): SomeType {
// ...
}
```
causes a problem.

Sboonny replied:
If we really want to discuss it, I have opened 2nd PR with buggy code, showing why it's great to use explicit types but it shouldn't be the default 

https://github.com/freeCodeCamp/freeCodeCamp/pull/49368

ShaunSHamilton commented on package.json:
```diff
@@ -18,58 +18,44 @@
+    "cypress:prd:watch": "pnpm run cypress -- open",
     "cypress:install": "cypress install && echo 'for use with ./cypress-install.js'",
     "cypress:install-build-tools": "sh ./cypress-install.sh",
     "predevelop": "npm-run-all create:*",
```
Is it still fine to use `npm-run-all`?

ojeytonwilliams replied:
Weirdly, yes: https://github.com/pnpm/pnpm/issues/413

ShaunSHamilton commented on package.json:
```diff
@@ -18,58 +18,44 @@
   "homepage": "https://github.com/freeCodeCamp/freeCodeCamp#readme",
   "author": "freeCodeCamp <team@freecodecamp.org>",
   "main": "none",
-  "workspaces": [
```
I cannot comment further. Line 9:

Should we remove/change `engines.npm` to `engines.pnpm`?

ojeytonwilliams replied:
Thanks, I'll do that.

ShaunSHamilton commented on package.json:
```diff
@@ -18,58 +18,44 @@
+    "build:curriculum": "cd ./curriculum && pnpm run build",
+    "build:server": "cd ./api-server && pnpm run build",
     "challenge-editor": "npm-run-all -p challenge-editor:*",
     "challenge-editor:client": "cd ./tools/challenge-editor/client && npm start",
     "challenge-editor:server": "cd ./tools/challenge-editor/api && npm start",
```
```suggestion
    "challenge-editor:client": "cd ./tools/challenge-editor/client && pnpm start",
    "challenge-editor:server": "cd ./tools/challenge-editor/api && pnpm start",
```

ShaunSHamilton commented on package.json:
```diff
@@ -82,24 +68,24 @@
+    "serve:client-ci": "cd ./client && pnpm run serve-ci",
     "start": "npm-run-all create:* -p develop:server serve:client",
     "start-ci": "npm-run-all create:* -p start:server serve:client-ci",
     "start:server": "cd ./api-server && npm start",
```
```suggestion
    "start:server": "cd ./api-server && pnpm start",
```

ShaunSHamilton commented on .github/workflows/node.js-tests.yml:
```diff
@@ -57,12 +62,12 @@ jobs:
       - name: Lint Source Files
         run: |
           echo npm version $(npm -v)
```
```suggestion
          echo pnpm version $(pnpm -v)
```

ojeytonwilliams committed:
fix: more npm run -> pnpm run

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ojeytonwilliams committed:
chore: update engines to use pnpm

ojeytonwilliams committed:
fix: pnpm less strict

pnpm engines doesn't respect ~, sadly

ojeytonwilliams committed:
fix: enable choice of (super)block for tests

Only 'nix machines for now.

raisedadead commented on api-server/package.json:
```diff
@@ -6,7 +6,7 @@
   "engines": {
     "node": ">=16",
-    "npm": ">=8"
+    "pnpm": ">=7.20.0"
```
Does is have to be this specific?

ojeytonwilliams replied:
Nope, I just failed at understanding semver (Shaun, please forgive me). '7' is all we need here, so we don't accidentally bump major versions.

ojeytonwilliams committed:
chore: pin pnpm to version 7

ojeytonwilliams committed:
fix: remove semver (the package, not the concept)

ojeytonwilliams committed:
chore: remove last npms

Except web + curriculum-server. I'll update them when I start work on
them again.

ojeytonwilliams committed:
fix: lockfile check to catch any package-locks

ojeytonwilliams committed:
fix(docs): revert npm -> pnpm in mobile docs

ojeytonwilliams committed:
fix(docs): update prerequisites

ojeytonwilliams commented:
That should be it. The one exception being the devops docs. @raisedadead assuming people are happy with this, I'll see what needs updating the pipelines + devops manuals.

ojeytonwilliams committed:
fix(action): install pnpm for upcoming tests

Sembauke commented on tools/challenge-auditor/index.ts:
```diff
@@ -10,9 +10,22 @@ config({ path: envPath });
-import { ChallengeNode } from '../../client/src/redux/prop-types';
 import { getAuditedSuperBlocks } from '../../config/superblock-order';
 
+// TODO: re-organise the types to a common 'types' folder that can be shared
+// between the workspaces so we don't have to declare ChallengeNode here and in
+// the client.
+
+// This cannot be imported from the client, without causing tsc to attempt to
+// compile the client (something it cannot do)
+type ChallengeNode = {
+  block: string;
+  dashedName: string;
+  superBlock: SuperBlocks;
+  id: string;
+  challengeType: number;
+};
+
```
Could you open an issue for this.

ojeytonwilliams committed:
chore: add nodemon to new api

ojeytonwilliams committed:
chore: generate pnpm-lock

ShaunSHamilton commented on curriculum/package.json:
```diff
@@ -25,7 +25,7 @@
     "update-step-titles": "cross-env CALLING_DIR=$INIT_CWD ts-node --project ../tsconfig.json ../tools/challenge-helper-scripts/update-step-titles",
-    "test": "ts-node ../node_modules/mocha/bin/mocha.js --delay --exit --reporter progress --bail",
+    "test": "ts-node ./node_modules/mocha/bin/mocha.js --delay --exit --reporter progress --bail",
     "test:full-output": "cross-env FULL_OUTPUT=true ts-node ../node_modules/mocha/bin/mocha.js --delay --reporter progress"
```
I realise I am the only one that ever uses this.

```suggestion
    "test:full-output": "cross-env FULL_OUTPUT=true ts-node ./node_modules/mocha/bin/mocha.js --delay --reporter progress"
```

ShaunSHamilton commented on docs/how-to-contribute-to-the-codebase.md:
```diff
@@ -205,12 +205,12 @@ A quick reference to the commands that you will need when working.
 | command                                                        | description                                                                         |
 | -------------------------------------------------------------- | ----------------------------------------------------------------------------------- |
 | `npm test`                                                     | Run all JS tests in the system, including client, server, lint and challenge tests. |
```
```suggestion
| `pnpm test`                                                     | Run all JS tests in the system, including client, server, lint and challenge tests. |
```

ShaunSHamilton commented on docs/how-to-work-on-practice-projects.md:
```diff
@@ -74,7 +74,7 @@ A one-off script that will automatically add the next step based on the last ste
 2. Run the following npm command:
```
Now, when we migrate to yarn, we will not need to update this

```suggestion
2. Run the following command:
```

ShaunSHamilton commented on docs/how-to-work-on-practice-projects.md:
```diff
@@ -89,7 +89,7 @@ A one-off script that automatically adds a specified number of steps. The challe
 2. Run the following npm command:
```
```suggestion
2. Run the following command:
```

ShaunSHamilton commented on docs/how-to-work-on-practice-projects.md:
```diff
@@ -104,7 +104,7 @@ A one-off script that automatically adds a new step at a specified position, inc
 2. Run the following npm command:
```
```suggestion
2. Run the following command:
```

ShaunSHamilton commented on docs/how-to-work-on-practice-projects.md:
```diff
@@ -119,7 +119,7 @@ A one-off script that deletes an existing step, decrementing all subsequent step
 2. Run the following npm command:
```
```suggestion
2. Run the following command:
```

ShaunSHamilton commented on docs/how-to-work-on-practice-projects.md:
```diff
@@ -132,5 +132,5 @@ A one-off script that automatically updates the frontmatter in a project's markd
 2. Run the following npm command:
```
```suggestion
2. Run the following command:
```

ShaunSHamilton commented on docs/how-to-work-on-the-component-library.md:
````diff
@@ -89,7 +89,7 @@ We use [React Testing Library](https://testing-library.com/docs/react-testing-li
+pnpm run test-ui-components
 ```
 
 ## Adding packages to the UI-Component library
````
One more npm below this line I am not sure why was left behind.

ShaunSHamilton reviewed (changes requested):
**Question**

For running scripts that have nothing to do with installing packages, does it make any difference using NPM vs PNPM?

**Note**

NPM recommends **not** using the `--prefix` flag: https://github.com/npm/cli/issues/1368
This is still used in CI, and in `web`.

As this does not really affect anything, I am happy with that being removed in a different PR.

Also, `docker` is littered with NPM commands. Am I correct in understanding we _currently_ do not use any of those configs?

ojeytonwilliams committed:
Merge branch 'main' into feat/pnpm

ojeytonwilliams committed:
fix: yet more npm -> pnpm

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ojeytonwilliams commented on .github/workflows/node.js-tests.yml:
```diff
@@ -35,12 +35,17 @@ jobs:
+          if [ $(find . -name 'package-lock.json' | grep -vc -e 'node_modules') -gt 0 ]
           then
-            echo -e 'Error: found too many lockfiles in the repository.\nWe use npm workspaces to manage packages so all dependencies should be added to the root package-lock.json'
+            echo -e 'Error: found too package-lock files in the repository.\nWe use pnpm workspaces to manage packages so all dependencies should be added via pnpm add'
```
```suggestion
            echo -e 'Error: found package-lock files in the repository.\nWe use pnpm workspaces to manage packages so all dependencies should be added via pnpm add'
```

ojeytonwilliams committed:
refactor: typo

ojeytonwilliams committed:
docs: pnpm add for ui-components

ojeytonwilliams commented:
> This is still used in CI, and in web.

If it's not failing in CI, I think we're okay. I'll update web when I get a chance.

ShaunSHamilton reviewed (approved):
LLTMAW 👍
